src/WebJobs.Script/Scale/HostPerformanceManager.cs

```diff
@@ -27,20 +41,139 @@ public HostPerformanceManager(IEnvironment environment, IOptions<HostHealthMonit
+        public virtual bool PerformanceCountersExceeded(Collection<string> exceededCounters = null, ILogger logger = null)
         {
             var counters = GetPerformanceCounters(logger);
             if (counters != null)
             {
-                return IsUnderHighLoad(counters, exceededCounters, _healthMonitorOptions.Value.CounterThreshold);
+                return PerformanceCounterThresholdsExceeded(counters, exceededCounters, _healthMonitorOptions.Value.CounterThreshold);
             }
 
             return false;
         }
 
-        internal static bool IsUnderHighLoad(ApplicationPerformanceCounters counters, Collection<string> exceededCounters = null, float threshold = HostHealthMonitorOptions.DefaultCounterThreshold)
+        /// <summary>
+        /// Check both sandbox enforced performance counters as well as process level thresholds
+        /// like CPU, memory, etc.
+        /// </summary>
+        public virtual async Task<bool> IsUnderHighLoad(Collection<string> exceededCounters = null, ILogger logger = null)
+        {
+            return PerformanceCountersExceeded(exceededCounters, logger) ||
+                await ProcessThresholdsExceeded(exceededCounters, logger);
+        }
+
+        public async Task<IActionResult> TryHandleHealthPingAsync(HttpRequest request, ILogger logger)
+        {
+            var healthPingEnabled = _environment.GetEnvironmentVariableOrDefault(EnvironmentSettingNames.HealthPingEnabled, "1");
+            if (healthPingEnabled.Equals("0"))
+            {
+                // disabled at app level
+                return null;
+            }
+
+            bool checkHealth = false;
+            var userAgent = request.GetHeaderValueOrDefault("User-Agent");
+            if (!string.IsNullOrEmpty(userAgent) &&
+                (userAgent.IndexOf(ScriptConstants.HttpScaleUserAgent, StringComparison.OrdinalIgnoreCase) != -1 ||
+                userAgent.IndexOf(ScriptConstants.ScaleControllerUserAgent, StringComparison.OrdinalIgnoreCase) != -1))
+            {
+                // for these user agents, we default to true
+                checkHealth = true;
+            }
+
+            var query = request.GetQueryCollectionAsDictionary();
+            if (query.TryGetValue(ScriptConstants.HealthCheckQueryParam, out string value))
+            {
+                // header overrides user agent
+                checkHealth = value.Equals("1");
+            }
+
+            if (checkHealth)
+            {
+                // check host + worker health
+                int statusCode = (int)HttpStatusCode.OK;
+                if (await IsUnderHighLoad(logger: logger))
+                {
+                    statusCode = 429;
+                }
+
+                return new StatusCodeResult(statusCode);
+            }
+
+            return null;
+        }
+
+        internal async Task<bool> ProcessThresholdsExceeded(Collection<string> exceededCounters = null, ILogger logger = null)
```

Comment:
This new process monitoring is the key for the new FE admin/host/ping scale ping handling. It enables us to return a status to influence scale out, that takes into account the current host and worker cpu load (and perhaps other metrics in the future).